compiler/src/lib.rs

```diff
@@ -1,90 +1,170 @@
+    //     Ok(x) => x,
+    //     Err(e) => return Err(locator.locate_error(e)),
+    // };
+
+    // TODO:
+    // if opts.optimize > 0 {
+    //     ast = ConstantOptimizer::new()
+    //         .fold_mod(ast)
+    //         .unwrap_or_else(|e| match e {});
+    // }
+    // let ast = locator.fold_mod(ast).unwrap_or_else(|e| match e {});
+}
+
+fn _compile(
+    source_code: SourceCode,
+    mode: core::Mode,
+    opts: CompileOpts,
+) -> Result<CodeObject, CompileError> {
+    let parsed = parser::parse(source_code.text, mode.into())
+        .map_err(|err| CompileError::from_ruff_parse_error(err, &source_code))?;
+    let ast = parsed.into_syntax();
+    compile::compile_top(ast, source_code, mode, opts).map_err(|e| e.into())
 }
 
 pub fn compile_symtable(
     source: &str,
     mode: Mode,
     source_path: &str,
 ) -> Result<symboltable::SymbolTable, CompileError> {
-    let mut locator = LinearLocator::new(source);
+    let source_code = SourceCode::new(source_path, source);
+    _compile_symtable(source_code, mode)
+}
+
+pub fn _compile_symtable(
+    source_code: SourceCode,
+    mode: Mode,
+) -> Result<symboltable::SymbolTable, CompileError> {
+    // let mut locator = LinearLocator::new(source);
     let res = match mode {
         Mode::Exec | Mode::Single | Mode::BlockExpr => {
-            let ast =
-                ast::Suite::parse(source, source_path).map_err(|e| locator.locate_error(e))?;
-            let ast = locator.fold(ast).unwrap();
-            symboltable::SymbolTable::scan_program(&ast)
+            let ast = ruff_python_parser::parse_module(source_code.text)
+                .map_err(|e| CompileError::from_ruff_parse_error(e, &source_code))?;
+            // let ast =
+            //     ast::Suite::parse(source, source_path).map_err(|e| locator.locate_error(e))?;
+            // let ast = locator.fold(ast).unwrap();
+            symboltable::SymbolTable::scan_program(&ast.into_syntax(), source_code.clone())
         }
         Mode::Eval => {
-            let expr =
-                ast::Expr::parse(source, source_path).map_err(|e| locator.locate_error(e))?;
-            let expr = locator.fold(expr).unwrap();
-            symboltable::SymbolTable::scan_expr(&expr)
+            // let expr =
+            //     ast::Expr::parse(source, source_path).map_err(|e| locator.locate_error(e))?;
+            // let expr = locator.fold(expr).unwrap();
+            let ast =
+                ruff_python_parser::parse(source_code.text, ruff_python_parser::Mode::Ipython)
+                    .map_err(|e| CompileError::from_ruff_parse_error(e, &source_code))?;
+            symboltable::SymbolTable::scan_expr(
+                &ast.into_syntax().expect_expression(),
+                source_code.clone(),
+            )
         }
     };
-    res.map_err(|e| e.into_codegen_error(source_path.to_owned()).into())
+    res.map_err(|e| e.into_codegen_error(source_code.path.to_owned()).into())
+}
+
```

Comment:
I couldn't find the root cause yet, but adding this test will make using debugger easy for the problem

```suggestion
#[test]
fn test_compile_phello() {
    let code = r#"
initialized = True

def main():
    print("Hello world!")

if __name__ == '__main__':
    main()
"#;
    let compiled = compile(&code, Mode::Exec, "<>", CompileOpts::default());
    dbg!(compiled.expect("compile error"));
}
```